### PR TITLE
Login with Google and Apple

### LINKED ISSUE #640: Add ContinueWithGoogle tapped action and button
## Description
Add the button, add the action, wire them up. The action should emit LogUserInWithThirdPartyEffect.

## Definition of done
Button is added
Action is added, reduced, tested

- [ ] The documentation is updated to reflect the changes done in this PR

### LINKED ISSUE #641: Add ContinueWithAppleTapped action and button
## Description
Add the button, add the action, wire them up. The action should emit LogUserInWithThirdPartyEffect.
## Definition of done
Button is added
Action is added, reduced, tested

- [ ] The documentation is updated to reflect the changes done in this PR

### LINKED ISSUE #642: Add LogUserInWithThirdPartyEffect - Google
## Description
Integrate Google Login. This should emit SetUser/LinkSso/SetError/MustAcceptTerms accordingly

## Definition of done
Effect is implemented
Effect can login the user with google
It detects when there’s no account and then emits MustAcceptTerms
Tests are added

- [ ] The documentation is updated to reflect the changes done in this PR

### LINKED ISSUE #643: Add LogUserInWithThirdPartyEffect - Apple
## Description
Integrate Apple Login. This should emit SetUser/LinkSso/SetError/MustAcceptTerms accordingly
## Definition of done
Effect is implemented
Effect can login the user with apple
It detects when there’s no account and then emits MustAcceptTerms
Tests are added

- [ ] The documentation is updated to reflect the changes done in this PR

### PR DESCRIPTION
**Pull request by simonrozsival**
### Summary
This PR adds Google and Apple login - it adds the buttons to the Onboarding screen, it adds the necessary third party auth and login actions, it obtains the auth tokens from Google or Apple, and it makes the necessary HTTP calls to V9 sign in the user (= to get the `User` object with the API token).

This PR does NOT attempt to make the onboarding screen look any good, it just adds the buttons and makes them work the way they're supposed to. The UI will be modified later in a separate PR (AFAIK there aren't even designs for this screen just yet).

This PR does NOT attempt signing up a new user account if the user doesn't have any account linked to that Google/Apple account.

### Technical considerations
- I added a `GoogleSignIn` pod
- I added an Apple auth entitlement

### What I'm still not sure about and what needs figuring out ASAP
The provisioning profile for the app should be changed to allow Apple auth if I understand it correctly. I'm not sure what exectly needs to be done though and if I have the rights to make such changes. It would be great if the reviewers with more experience than me helped me figure it out 😄 ❤️ 

### Why is it done like this and not some other way?
I didn't modify the state class at all because so far it seems to me that it's not necessary. That might change though when we start implementing the sign up part, which needs to remember some state before the user agrees to the Terms of Service and Privacy Policy. Currently it's just one action leading to another action.

### Relationships
Closes #640
Closes #641
Closes #642
Closes #643 

### Pull Request checklist
- [x] The code does not hinder the app's performance
- [x] The code does not introduce regressions
- [x] The code in this PR is tested thoroughly enough that changing the behavior to an incorrect one breaks tests

### Pointers
- I added a `GoogleSignIn` pod and I had to update the other pods (I removed the lock file and reinstalled) because otherwise the app won't build for some reason 🤷 Please check that I didn't mess things up and if needed, tell me how to achieve the same result without modifying the other pods as well

### Review template

- [ ] The code is thoroughly tested in well written and resilient manner
- [ ] The code is performant and leads to an optimal user experience
- [ ] The code is testable and does not cut corners for the sake of delivering
- [ ] The code considers older platforms well and will work in older devices just like it will in new ones

- [ ] The code considers light and dark theme
- [ ] The code considers RTL and does not mention left/right
- [ ] The code considers i18n
  - [ ] There are no hardcoded strings in the UI
  - [ ] All strings are included in every language we support OR there's an issue requesting new translations
- [ ] There are no hard-coded numbers referring to dimensions or font sizes
- [ ] The code considers UI elements that are similar and extracts their similarities into styles to avoid duplication

## Merge permissions
🦑

### CODE REVIEW

**Comment by sofokli1** [comment]
did you do `pod install` or `pod update` after adding the new pod in the podfile?

**Comment by sofokli1** [comment]
as for the apple sign in provisioning profile changes needed, @juanlaube is your man since he had do that already for the old app.

**Comment by sofokli1** [comment]
i do not see a sign in with google button. how can I test this in practice?

**Comment by simonrozsival** [comment]
> i do not see a sign in with google button. how can I test this in practice?

@sofokli1 yeah it's because the google client ID is not set up when in debug mode - you need to edit https://github.com/toggl/ios/pull/704/files#diff-3d80ad5430d58e07f07ea4c9c9f65153R45 and put the client ID string there directly (I can tell you more in Slack).

**Comment by simonrozsival** [comment]
@juanlaube kindly updated the provisioning profile and now Apple Sign In works well on an actual device. It doesn't work in the simulator for some reason (at least you can test the "Oops!" alert). The only issue is that we don't have the initial sync set up yet and therefore the app crashes as soon as you're logged in. Check the log to see what's going on.

**Comment by juanlaube** [comment]
@simonrozsival I got it working on the simulator, but I had to login with my iCloud account first, I had this in the logs after that: `Action: ThirdPartyAuthSucceeded(apple("<redacted>"), Juan Laube)`

File: Podfile.lock [hunk 1]
```diff
@@ -184,36 +186,35 @@ SPEC REPOS:
...
   RxDataSources: efee07fa4de48477eca0a4611e6d11e2da9c1114
   RxRelay: d77f7d771495f43c556cbc43eebd1bb54d01e8e9
   RxSwift: 81470a2074fa8780320ea5fe4102807cb7118178
   RxTest: 711632d5644dffbeb62c936a521b5b008a1e1faa
 
-PODFILE CHECKSUM: 38319e32fda31d7b189cf52fd7cecfde3364d3d6
+PODFILE CHECKSUM: fdc7578ee50d011c649544f4e0221c8d98dd799b
 
-COCOAPODS: 1.9.1
+COCOAPODS: 1.10.0.beta.1
```
Comment (dshr) [inline Podfile.lock:220]: Why are you running the beta ver?

**Review by dshr — CHANGES_REQUESTED** [review]
All good! I tried my luck with de-updating the Pods, but tbh I think it's not a big deal, and they're ok to be updated. So all that needs to be done is some 💄 stuff.

File: Podfile.lock (hunk 1 above)
Comment (simonrozsival) [reply Podfile.lock]: I actually didn't notice. For some reason my `pod` binary was a beta one. I uninstalled it, remove `Podfile.lock`, and reinstalled pods.

File: Modules/Assets/Assets/Extensions/Strings.swift [hunk 2]
```diff
@@ -80,6 +80,8 @@ public struct Strings {
         public static var passwordResetHelpCTA: String { "passwordResetHelpCTA".localized }
         public static var passwordResetSuccessTitle: String { "passwordResetSuccessTitle".localized }
         public static var passwordResetSuccessText: String { "passwordResetSuccessText".localized }
+        public static var oops: String { "oops".localized }
+        public static var loginGeneralError: String { "loginGeneralError".localized }
```
[the remark this answers is not in the document]
Comment (simonrozsival) [reply Modules/Assets/Assets/Extensions/Strings.swift:84]: I guess that's caused by merge stuff from other PRs 😄

File: Modules/Assets/Assets/fr.lproj/Localizable.strings [hunk 3]
```diff
@@ -89,7 +89,6 @@
 "cancel" = "Annuler";
 "discard" = "Jeter";
 "continueEditing" = "Continuer l'édition";
-"oops" = "Oops!";
```
[the remark this answers is not in the document]
Comment (simonrozsival) [reply Modules/Assets/Assets/fr.lproj/Localizable.strings:92]: Bad merge I guess? 😆

**Comment by simonrozsival** [comment]
I don't understand the output of the linter (:x:). Running `swiftlint` locally doesn't show any violations and some of the messages seem unrelated to this PR:

```
/Users/vagrant/git/Modules/Settings/Settings/Common/View/SettingsViewController.swift:121:42:
warning: Empty Enum Arguments Violation:
Arguments can be omitted when matching enums with associated types if they are not used. (empty_enum_arguments)
```
-- I didn't change anything in `SettingsViewController` or anything related to the line 121 (`} else if case .error(_) = result {`)

```
/Users/vagrant/git/Modules/Calendar/Calendar/CalendarDay/CalendarLayoutCalculator.swift:48:35:
warning: Empty Enum Arguments Violation:
Arguments can be omitted when matching enums with associated types if they are not used. (empty_enum_arguments)
```
-- same story here....

```
/Users/vagrant/git/Modules/Timer/Timer/TimeEntriesLog/TimeEntriesSelector.swift:26:5:
warning: Closure Parameter Position Violation:
Closure parameters should be on the same line as opening brace. (closure_parameter_position)
```
-- no idea how that got into `master` and now causes problems here 🤷 

The linter check marks 12 violations, all either `closure_parameter_position` or `empty_enum_arguments`. Is there some reasonable explanation for this weird behavior?

**Comment by dshr** [comment]
I think this is related to swiftlint version mismatches, this changing the default set of rules. My local version is `0.39.1`, while Bitrise is running `0.4.2`. I'm gonna try updating to that and will update here.

**Comment by dshr** [comment]
So, 0.4.2 is the version of the swiftlint integration package, not the actual binary. I've updated to the latest version on Homebrew, which is 0.40.0, but I still don't get those errors

**Review by dshr — APPROVED** [review]

**Pull request merged**